### PR TITLE
enhancement(logplex source): annotate logs with query parameters

### LINKED ISSUE #4732: Support annotating logs with URL parameters for the logplex source
Discussion in discord: https://discordapp.com/channels/742820443487993987/746070591097798688/769706197547876362

In its current form, it seems like the `logplex` source is most useful when used as the logdrain for a single Heroku app because it isn't possible to differentiate the logs by Heroku app.

Example set of Heroku logs for a single app:

```
2020-10-24T08:40:12.187823+00:00 app[web.1]: * Listening on tcp://0.0.0.0:38071
2020-10-24T08:40:12.188274+00:00 app[web.1]: Use Ctrl-C to stop
2020-10-24T08:40:12.621499+00:00 heroku[web.1]: State changed from starting to up
2020-10-24T08:40:13.137687+00:00 heroku[router]: at=info method=GET path="/robots.txt" host=adoptadrainsf-staging.herokuapp.com request_id=a05a4bad-43b5-4bbd-9138-4e7a32d4d0c2 fwd="157.55.39.39" dyno=web.1 connect=0ms service=8ms status=200 bytes=333 protocol=http
2020-10-24T08:40:31.705732+00:00 heroku[router]: at=info method=GET path="/sitemap.xml" host=adoptadrainsf-staging.herokuapp.com request_id=db5e6e9b-729a-4cf7-9976-857953979ddf fwd="52.247.221.252" dyno=web.1 connect=3ms service=101ms status=200 bytes=686 protocol=http
2020-10-24T09:15:37.291335+00:00 heroku[web.1]: Idling
```

Here we can see that Heroku uses the app name field to differentiate between runtime logs, router logs, and app logs.

The proposed idea, which I also like, is to allow parsing out URL parameters for that source and adding them as log fields. This would let users set their logdrain to things like:

```
https://url.for.vector:443?app=myapplicationname
```

The `logplex` source could then use this to annotate all of the logs with `app`.

We could also consider this for other HTTP-based sources. We have some precedent of doing so with headers for the generic `http` source.

### PR DESCRIPTION
**Pull request by CGA1123**
Allow annotation of logs via query parameters for http + logplex source,
in a similar fashion as is currently supported for HTTP headers.

This is particularly useful with the logplex source as it allows for a
single vector instance to be used as a drain for multiple Heroku
applications.

Closes #4732

Signed-off-by: Christian Gregg <christian@bissy.io>

### CODE REVIEW

File: src/sources/http.rs
Comment (CGA1123) [inline src/sources/http.rs:137]: This is duplicated exactly in `logplex`, maybe could be pulled into `util/http.rs`?

File: src/sources/logplex.rs
Comment (jszwedko) [inline src/sources/logplex.rs:27]: Since this will be pulled out as the public facing user configuration, could we call it `query_parameters`?

File: src/sources/logplex.rs
Comment (jszwedko) [inline src/sources/logplex.rs:133]: I'm seeing that this matches `add_headers`, but I'm wondering if we shouldn't map `None` to `event::Value::Null` rather than the empty string. That would help distinguish parameters that are missing from those that exist, but are empty. I think `add_headers` may have predated `event::Value::Null`; I wouldn't mind seeing that updated too.

File: src/sources/http.rs
Comment (jszwedko) [reply src/sources/http.rs:137]: :+1: I think that's a good idea. Maybe `src/sources/util/http.rs` since I think only sources would make use of it.

**Review by jszwedko — COMMENTED** [review]
Thanks for this @CGA1123 !

Just one note about mapping non-existent parameters, but I think the rest looks good.

We'll want to add this to the docs too. If you feel like taking a shot, you can find them in `docs/`, otherwise we can add them once this is merged.

**Comment by CGA1123** [comment]
Hi @jszwedko thanks for the review, I think I have addressed your comments!
I took a shot at updating the docs as well 😄 

CI seems to have hanged on the mac build though, not sure why...

**Review by jszwedko — APPROVED** [review]
Looks great! Thanks @CGA1123 . I kicked CI to run it again and will merge once it is green.

**Comment by CGA1123** [comment]
Yay! Thanks @jszwedko 😄 

Will be trying this out in the coming days on our apps at work I think 🚀

**Pull request merged**
